File: 4_7/p054.py
def new_list():
    return [[0]*2 for _ in range(2)]

def multiple(a, b, m):
    ans = new_list()
    for i in range(2):
        for j in range(2):
            for k in range(2):
                ans[i][j] += a[i][k]*b[k][j]
            ans[i][j] = ans[i][j] % m
    return ans

def bin_pow(r, e, m):
    res = r[:]
    flg = False
    for i in range(65):
        if e & 1 << i:
            if not flg:
                res = r[:]
                flg = True
            else:
                res = multiple(res, r, m)
        r = multiple(r, r, m)
    return res

n = int(input())
A = new_list()
A[0][0], A[0][1], A[1][0] = 1, 1, 1
M = 10 ** 9
ans = bin_pow(A, n-1, M)
ans = (ans[1][0]+ans[1][1]) % M
print(ans)

File: 4_7/p054.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var scan = bufio.NewScanner(os.Stdin)

const (
	initBufSize = 10000
	maxBufSize = 10000000
	M = 1000000000
)

func init() {
	buf := make([]byte, initBufSize)
	scan.Buffer(buf, maxBufSize)
	scan.Split(bufio.ScanWords)
}

func main() {
	n := NumStdin()
	res := pow(n)
	ans := 0
	for _, v := range res[1] {
		ans += v
		ans %= M
	}
	fmt.Println(ans)
}

func NumStdin() int {
	scan.Scan()
	num, err := strconv.Atoi(scan.Text())
	if err != nil {
		panic(err)
	}
	return num
}

func pow(n int) [2][2]int {
	n--
	A := [2][2]int{{1, 1}, {1, 0}}
	ret := [2][2]int{}

	flg := true
	for i := 0; i < 60; i++ {
		if (n & (1 << i)) != 0 {
			if flg {
				ret = A
				flg = false
			}else{
				ret = multi(ret, A)
			}
		}
		A = multi(A, A)
	}
	return ret
}

func multi(a, b [2][2]int) [2][2]int {
	ret := [2][2]int{}
	for i := 0; i < 2; i++ {
		for j := 0; j < 2; j++ {
			for k := 0; k < 2; k++ {
				ret[i][j] += a[i][k] * b[k][j]
				ret[i][j] %= M
			}
		}
	}
	return ret
}
